packages/tests/server/react/useContext.test.tsx

```diff
@@ -289,6 +290,87 @@ test('invalidate', async () => {
   expect(stableProxySpy).toHaveBeenCalledTimes(1);
 });
 
+test('invalidate procedure for both query and infinite', async () => {
+  const { proxy, App } = ctx;
+  const invalidateQuerySpy = jest.fn();
+  const invalidateInfiniteSpy = jest.fn();
+
+  function MyComponent() {
+    const allPostsList = proxy.post.list.useQuery(
+      { cursor: undefined },
+      {
+        onSuccess: invalidateQuerySpy,
+      },
+    );
+    const allPostsListInfinite = proxy.post.list.useInfiniteQuery(
+      { cursor: undefined },
+      { onSuccess: invalidateInfiniteSpy },
+    );
+
+    const utils = proxy.useContext();
+
+    return (
+      <>
+        <button
+          data-testid="invalidate-button"
+          onClick={() => {
+            utils.post.list.invalidate();
+          }}
+        />
+        <div data-testid="list-status">
+          {allPostsList.isFetching || allPostsListInfinite.isFetching
+            ? 'fetching'
+            : 'done'}
+        </div>
+        <div>
+          {allPostsListInfinite.data?.pages.map((page) => {
+            return page.map((post) => {
+              return <div key={post.id}>{post.text}</div>;
+            });
+          })}
+        </div>
+      </>
+    );
+  }
+
+  const utils = render(
+    <App>
+      <MyComponent />
+    </App>,
+  );
+
+  await waitFor(() => {
+    expect(utils.container).toHaveTextContent('done');
+  });
+
+  await waitFor(() => {
+    expect(utils.container).toHaveTextContent('new post');
+  });
+
+  // I've tried everything to avoid this but for some reason, even though
+  // infinite query is successful and data is on the component it still shows
+  // no calls to the proxy passed to "onSuccess" unless i do this wait ... even
+  // await new Promise(process.nextTick); doesn't work 🤷🤯
+  await waitMs(1);
```

Comment:
I'm not happy with this but I cannot for the life of me find out why this test will not work without the wait?! ... 

It seems like something weird is happening with the calling of the `onSuccess` callback on `.useInfiniteQuery()`

@TkDodo - Am I doing anything stupid here you can see straight off?? 

If we don't want to introduce this, we could remove this leaf node test and just keep the router test also introduced in this PR ... but they are slightly different routes, so I ideally wanted to test both in case they get implemented differently in the future...

Reply:
a bit strange because it works in TanStack/query:

https://github.com/TanStack/query/blob/933cd410ae0c8e69f9375eac2d5fb38cc940e6eb/packages/react-query/src/__tests__/useQuery.test.tsx#L434-L458

However, I can't see how the queryFn is implemented in the test. There is lots of batching going on inside RQ, so if you queryFn is synchronous, it could be that there is a render that has `isFetching:false` immediately while _not_ having called `onSuccess`. As you can see, in our tests, we made the `queryFn` async and waited a bit in there.

Reply:
The query function being used in this `infiniteQuery` is horribly simple and not async / how an infinite query function would normally work so maybe it is something to do with that! 

```ts
....
type Post = {
  id: number;
  text: string;
};

const defaultPost = { id: 0, text: 'new post' };
const ctx = konn()
  .beforeEach(() => {
    const t = initTRPC.create({
      errorFormatter({ shape }) {
        return {
          ...shape,
          data: {
            ...shape.data,
            foo: 'bar' as const,
          },
        };
      },
    });

    const posts: Post[] = [defaultPost];

    const appRouter = t.router({
      post: t.router({
        list: t.procedure
          .input(
            z.object({
              cursor: z.string().optional(),
            }),
          )
          .query(() => posts),
```

I'll have a quick look if making the query function async fixes it but seems odd! 

Thanks for the rapid response though 🙂🤯.

Reply:
Making the query function async and adding a 10ms wait still didn't seems to solve it without waiting specifically in the test 🤔

It must be something to do with the order that things happen between `isFetching` being set to false and the `onSuccess` callback being called 🤷 ... Don't understand why `await new Promise(process.nextTick);` doesn't work though when a 1ms wait does!